README.md
# paratranz
本文档介绍 ParaTranz.cn 平台的 API
## 获取 Token
首先需要获取API Token，可以通过点击个人资料页面左侧的小钥匙图标获取API Token， 调用 API 时将 Token 直接放在请求头的 Authorization 中即可。

cURL 使用示例:

    $ curl --header \"Authorization: XXXXXXXXX\" https://paratranz.cn/api/projects

## 频率限制
每分钟120次调用，超出后将会降速（增加延迟），1分钟后自动恢复。
## 错误处理
API 返回的错误格式如下

    {
      \"message\": \"ERROR MESSAGE\", // 错误消息
      \"code\": 10000 // 5位错误代码，注意与下面的HTTP状态码区分，部分接口不返回
    }

HTTP状态码有以下几种类型
  * 400 - 调用参数错误
  * 401 - Token 错误或过期
  * 403 - 没有相关权限
  * 404 - 资源不存在
  * 405 - 没有相关HTTP方法，一般为调用方法错误
  * 429 - 调用过于频繁，具体频率限制请看上一节
  * 500 - 服务器错误，一般会提供具体出错的位置，请发送给站长方便定位问题
  * 502 - 服务器无响应，部分用户被墙时可能会遇到
  * 503 - 服务不可用
  * 504 - 服务超时，访问量大时会出现

## SDK 及 JSON 格式的 API 文档
本文档遵循 OpenAPI 规范，[点击此处](https://paratranz.cn/api-docs?raw=1) 获取 JSON 格式的文档，您可以使用 [OpenAPI Generator](https://github.com/OpenAPITools/openapi-generator) 生成各种语言的 SDK
## 更新历史
  * v0.3.5 - 2023.12.25 修复部分 schema 问题
  * v0.3.4 - 2023.12.23 增加API调用频率限制说明
  * v0.3.3 - 2023.08.11 增加项目成员相关接口说明
  * v0.3.2 - 2022.11.04 增加文件翻译相关接口说明
  * v0.3.1 - 2022.10.16 修改 tag 及 schema 以便生成 sdk
  * v0.3.0 - 2022.10.16 增加术语历史记录接口说明，调整历史记录接口字段; 增加文档中 operationId 定义;
                        修复项目信息相关接口格式定义; 增加 JSON 格式文档入口
  * v0.2.1 - 2022.07.23 增加成员贡献接口文档; 完善列表接口数据结构
  * v0.2.0 - 2022.06.15 增加讨论及私信相关接口文档
  * v0.1.3 - 2022.03.10 增加历史记录相关接口文档
  * v0.1.2 - 2022.02.07 完善词条搜索接口 query 参数说明
  * v0.1.1 - 2022.01.17 增加文件历史相关接口文档
  * v0.1.0 - 2022.01.12 初次发布


This Python package is automatically generated by the [OpenAPI Generator](https://openapi-generator.tech) project:

- API version: 0.3.4
- Package version: 3.0.4
- Build package: org.openapitools.codegen.languages.PythonClientCodegen

## Requirements.

Python 3.7+

## Installation & Usage
### pip install

If the python package is hosted on a repository, you can install directly using:

```sh
pip install git+https://github.com/MuXiu1997/ParaTranz-SDK-python.git
```
(you may need to run `pip` with root permission: `sudo pip install git+https://github.com/MuXiu1997/ParaTranz-SDK-python.git`)

Then import the package:
```python
import paratranz_client
```

### Setuptools

Install via [Setuptools](http://pypi.python.org/pypi/setuptools).

```sh
python setup.py install --user
```
(or `sudo python setup.py install` to install the package for all users)

Then import the package:
```python
import paratranz_client
```

### Tests

Execute `pytest` to run the tests.

## Getting Started

Please follow the [installation procedure](#installation--usage) and then run the following:

```python

import time
import paratranz_client
from paratranz_client.rest import ApiException
from pprint import pprint

# Defining the host is optional and defaults to https://paratranz.cn/api
# See configuration.py for a list of all supported configuration parameters.
configuration = paratranz_client.Configuration(
    host = "https://paratranz.cn/api"
)

# The client must configure the authentication and authorization parameters
# in accordance with the API server security policy.
# Examples for each auth method are provided below, use the example that
# satisfies your auth use case.

# Configure API key authorization: Token
configuration.api_key['Token'] = os.environ["API_KEY"]

# Uncomment below to setup prefix (e.g. Bearer) for API key, if needed
# configuration.api_key_prefix['Token'] = 'Bearer'


# Enter a context with an instance of the API client
async with paratranz_client.ApiClient(configuration) as api_client:
    # Create an instance of the API class
    api_instance = paratranz_client.ArtifactsApi(api_client)
    project_id = 56 # int | 项目ID

    try:
        # 下载
        await api_instance.download_artifact(project_id)
    except ApiException as e:
        print("Exception when calling ArtifactsApi->download_artifact: %s\n" % e)

```

## Documentation for API Endpoints

All URIs are relative to *https://paratranz.cn/api*

Class | Method | HTTP request | Description
------------ | ------------- | ------------- | -------------
*ArtifactsApi* | [**download_artifact**](docs/ArtifactsApi.md#download_artifact) | **GET** /projects/{projectId}/artifacts/download | 下载
*ArtifactsApi* | [**generate_artifact**](docs/ArtifactsApi.md#generate_artifact) | **POST** /projects/{projectId}/artifacts | 触发导出
*ArtifactsApi* | [**get_artifact**](docs/ArtifactsApi.md#get_artifact) | **GET** /projects/{projectId}/artifacts | 导出结果
*FilesApi* | [**create_file**](docs/FilesApi.md#create_file) | **POST** /projects/{projectId}/files | 上传文件
*FilesApi* | [**delete_file**](docs/FilesApi.md#delete_file) | **DELETE** /projects/{projectId}/files/{fileId} | 删除文件
*FilesApi* | [**get_file**](docs/FilesApi.md#get_file) | **GET** /projects/{projectId}/files/{fileId} | 文件信息
*FilesApi* | [**get_file_translation**](docs/FilesApi.md#get_file_translation) | **GET** /projects/{projectId}/files/{fileId}/translation | 文件翻译
*FilesApi* | [**get_files**](docs/FilesApi.md#get_files) | **GET** /projects/{projectId}/files | 文件列表
*FilesApi* | [**save_file**](docs/FilesApi.md#save_file) | **PUT** /projects/{projectId}/files/{fileId} | 修改文件
*FilesApi* | [**update_file**](docs/FilesApi.md#update_file) | **POST** /projects/{projectId}/files/{fileId} | 更新文件
*FilesApi* | [**update_file_translation**](docs/FilesApi.md#update_file_translation) | **POST** /projects/{projectId}/files/{fileId}/translation | 更新文件翻译
*HistoryApi* | [**get_file_revisions**](docs/HistoryApi.md#get_file_revisions) | **GET** /projects/{projectId}/files/revisions | 文件历史
*HistoryApi* | [**get_history**](docs/HistoryApi.md#get_history) | **GET** /history | 获取历史记录
*HistoryApi* | [**get_term_history**](docs/HistoryApi.md#get_term_history) | **GET** /projects/{projectId}/terms/{termId}/history | 术语历史
*IssuesApi* | [**create_issue**](docs/IssuesApi.md#create_issue) | **POST** /projects/{projectId}/issues | 发起讨论
*IssuesApi* | [**delete_issue**](docs/IssuesApi.md#delete_issue) | **DELETE** /projects/{projectId}/issues/{issueId} | 删除讨论
*IssuesApi* | [**get_issue**](docs/IssuesApi.md#get_issue) | **GET** /projects/{projectId}/issues/{issueId} | 讨论信息
*IssuesApi* | [**get_issues**](docs/IssuesApi.md#get_issues) | **GET** /projects/{projectId}/issues | 讨论列表
*IssuesApi* | [**operate_issue**](docs/IssuesApi.md#operate_issue) | **POST** /projects/{projectId}/issues/{issueId} | 操作讨论
*IssuesApi* | [**save_issue**](docs/IssuesApi.md#save_issue) | **PUT** /projects/{projectId}/issues/{issueId} | 修改讨论
*MailsApi* | [**create_mail**](docs/MailsApi.md#create_mail) | **POST** /mails | 发送私信
*MailsApi* | [**get_conversation**](docs/MailsApi.md#get_conversation) | **GET** /mails/conversations/{userId} | 用户对话
*MailsApi* | [**get_mails**](docs/MailsApi.md#get_mails) | **GET** /mails | 私信列表
*MembersApi* | [**create_member**](docs/MembersApi.md#create_member) | **POST** /projects/{projectId}/members | 创建成员
*MembersApi* | [**delete_member**](docs/MembersApi.md#delete_member) | **DELETE** /projects/{projectId}/members/{memberId} | 删除成员
*MembersApi* | [**edit_member**](docs/MembersApi.md#edit_member) | **PUT** /projects/{projectId}/members/{memberId} | 修改成员
*MembersApi* | [**get_members**](docs/MembersApi.md#get_members) | **GET** /projects/{projectId}/members | 成员列表
*ProjectsApi* | [**create_project**](docs/ProjectsApi.md#create_project) | **POST** /projects | 创建项目
*ProjectsApi* | [**delete_project**](docs/ProjectsApi.md#delete_project) | **DELETE** /projects/{projectId} | 删除项目
*ProjectsApi* | [**get_project**](docs/ProjectsApi.md#get_project) | **GET** /projects/{projectId} | 项目信息
*ProjectsApi* | [**get_projects**](docs/ProjectsApi.md#get_projects) | **GET** /projects | 项目列表
*ProjectsApi* | [**save_project**](docs/ProjectsApi.md#save_project) | **PUT** /projects/{projectId} | 更新项目
*ScoresApi* | [**get_scores**](docs/ScoresApi.md#get_scores) | **GET** /projects/{projectId}/scores | 成员贡献
*StringsApi* | [**create_string**](docs/StringsApi.md#create_string) | **POST** /projects/{projectId}/strings | 创建词条
*StringsApi* | [**delete_string**](docs/StringsApi.md#delete_string) | **DELETE** /projects/{projectId}/strings/{stringId} | 删除词条
*StringsApi* | [**get_string**](docs/StringsApi.md#get_string) | **GET** /projects/{projectId}/strings/{stringId} | 获取词条
*StringsApi* | [**get_strings**](docs/StringsApi.md#get_strings) | **GET** /projects/{projectId}/strings | 词条列表
*StringsApi* | [**save_string**](docs/StringsApi.md#save_string) | **PUT** /projects/{projectId}/strings/{stringId} | 更新词条
*TermsApi* | [**create_term**](docs/TermsApi.md#create_term) | **POST** /projects/{projectId}/terms | 创建术语
*TermsApi* | [**delete_term**](docs/TermsApi.md#delete_term) | **DELETE** /projects/{projectId}/terms/{termId} | 删除术语
*TermsApi* | [**get_term**](docs/TermsApi.md#get_term) | **GET** /projects/{projectId}/terms/{termId} | 术语信息
*TermsApi* | [**get_terms**](docs/TermsApi.md#get_terms) | **GET** /projects/{projectId}/terms | 术语列表
*TermsApi* | [**import_terms**](docs/TermsApi.md#import_terms) | **PUT** /projects/{projectId}/terms | 批量导入术语
*TermsApi* | [**save_term**](docs/TermsApi.md#save_term) | **PUT** /projects/{projectId}/terms/{termId} | 修改术语
*UsersApi* | [**get_user**](docs/UsersApi.md#get_user) | **GET** /users/{userId} | 获取用户信息
*UsersApi* | [**save_user**](docs/UsersApi.md#save_user) | **PUT** /users/{userId} | 更新用户


## Documentation For Models

 - [Artifact](docs/Artifact.md)
 - [Comment](docs/Comment.md)
 - [CreateFile200Response](docs/CreateFile200Response.md)
 - [CreateIssueRequest](docs/CreateIssueRequest.md)
 - [EditMemberRequest](docs/EditMemberRequest.md)
 - [File](docs/File.md)
 - [GetFileRevisions200Response](docs/GetFileRevisions200Response.md)
 - [GetHistory200Response](docs/GetHistory200Response.md)
 - [GetIssue200Response](docs/GetIssue200Response.md)
 - [GetIssues200Response](docs/GetIssues200Response.md)
 - [GetIssues200ResponseResultsInner](docs/GetIssues200ResponseResultsInner.md)
 - [GetMails200Response](docs/GetMails200Response.md)
 - [GetProjects200ResponseInner](docs/GetProjects200ResponseInner.md)
 - [GetScores200Response](docs/GetScores200Response.md)
 - [GetStrings200Response](docs/GetStrings200Response.md)
 - [GetTermHistory200Response](docs/GetTermHistory200Response.md)
 - [GetTerms200Response](docs/GetTerms200Response.md)
 - [History](docs/History.md)
 - [HistoryTarget](docs/HistoryTarget.md)
 - [ImportTerms200Response](docs/ImportTerms200Response.md)
 - [Issue](docs/Issue.md)
 - [Job](docs/Job.md)
 - [Mail](docs/Mail.md)
 - [MailFromUser](docs/MailFromUser.md)
 - [Member](docs/Member.md)
 - [OperateIssueRequest](docs/OperateIssueRequest.md)
 - [Permission](docs/Permission.md)
 - [Project](docs/Project.md)
 - [Reply](docs/Reply.md)
 - [Revision](docs/Revision.md)
 - [SaveFileRequest](docs/SaveFileRequest.md)
 - [SaveUserRequest](docs/SaveUserRequest.md)
 - [Score](docs/Score.md)
 - [Stage](docs/Stage.md)
 - [StringItem](docs/StringItem.md)
 - [Term](docs/Term.md)
 - [TermHistory](docs/TermHistory.md)
 - [TermHistoryTarget](docs/TermHistoryTarget.md)
 - [User](docs/User.md)


<a id="documentation-for-authorization"></a>
## Documentation For Authorization


Authentication schemes defined for the API:
<a id="Token"></a>
### Token

- **Type**: API key
- **API key parameter name**: Authorization
- **Location**: HTTP header


## Author

master@mail.paratranz.com




docs/History.md
# History

词条历史记录

## Properties

Name | Type | Description | Notes
------------ | ------------- | ------------- | -------------
**id** | **int** |  | [optional] 
**created_at** | **datetime** |  | [optional] [readonly] 
**field** | **str** | 修改的字段 | [optional] 
**uid** | **int** | 用户ID | [optional] [readonly] 
**user** | **object** | 操作者用户详情 | [optional] [readonly] 
**tid** | **int** | 词条ID | [optional] 
**var_from** | **str** | 修改前的值 | [optional] 
**to** | **str** | 修改后的值 | [optional] 
**target** | [**HistoryTarget**](HistoryTarget.md) |  | [optional] 
**operation** | **str** | 历史记录操作类型 | [optional] 

## Example

```python
from paratranz_client.models.history import History

# TODO update the JSON string below
json = "{}"
# create an instance of History from a JSON string
history_instance = History.from_json(json)
# print the JSON string representation of the object
print History.to_json()

# convert the object into a dict
history_dict = history_instance.to_dict()
# create an instance of History from a dict
history_form_dict = history.from_dict(history_dict)
```
[[Back to Model list]](../README.md#documentation-for-models) [[Back to API list]](../README.md#documentation-for-api-endpoints) [[Back to README]](../README.md)




paratranz_client/rest.py
# coding: utf-8

"""
    ParaTranz OpenAPI 文档

    本文档介绍 ParaTranz.cn 平台的 API ## 获取 Token 首先需要获取API Token，可以通过点击个人资料页面左侧的小钥匙图标获取API Token， 调用 API 时将 Token 直接放在请求头的 Authorization 中即可。  cURL 使用示例:      $ curl --header \"Authorization: XXXXXXXXX\" https://paratranz.cn/api/projects  ## 频率限制 每分钟120次调用，超出后将会降速（增加延迟），1分钟后自动恢复。 ## 错误处理 API 返回的错误格式如下      {       \"message\": \"ERROR MESSAGE\", // 错误消息       \"code\": 10000 // 5位错误代码，注意与下面的HTTP状态码区分，部分接口不返回     }  HTTP状态码有以下几种类型   * 400 - 调用参数错误   * 401 - Token 错误或过期   * 403 - 没有相关权限   * 404 - 资源不存在   * 405 - 没有相关HTTP方法，一般为调用方法错误   * 429 - 调用过于频繁，具体频率限制请看上一节   * 500 - 服务器错误，一般会提供具体出错的位置，请发送给站长方便定位问题   * 502 - 服务器无响应，部分用户被墙时可能会遇到   * 503 - 服务不可用   * 504 - 服务超时，访问量大时会出现  ## SDK 及 JSON 格式的 API 文档 本文档遵循 OpenAPI 规范，[点击此处](https://paratranz.cn/api-docs?raw=1) 获取 JSON 格式的文档，您可以使用 [OpenAPI Generator](https://github.com/OpenAPITools/openapi-generator) 生成各种语言的 SDK ## 更新历史   * v0.3.5 - 2023.12.25 修复部分schema格式问题   * v0.3.4 - 2023.12.23 增加API调用频率限制说明   * v0.3.3 - 2023.08.11 增加项目成员相关接口说明   * v0.3.2 - 2022.11.04 增加文件翻译相关接口说明   * v0.3.1 - 2022.10.16 修改 tag 及 schema 以便生成 sdk   * v0.3.0 - 2022.10.16 增加术语历史记录接口说明，调整历史记录接口字段; 增加文档中 operationId 定义;                         修复项目信息相关接口格式定义; 增加 JSON 格式文档入口   * v0.2.1 - 2022.07.23 增加成员贡献接口文档; 完善列表接口数据结构   * v0.2.0 - 2022.06.15 增加讨论及私信相关接口文档   * v0.1.3 - 2022.03.10 增加历史记录相关接口文档   * v0.1.2 - 2022.02.07 完善词条搜索接口 query 参数说明   * v0.1.1 - 2022.01.17 增加文件历史相关接口文档   * v0.1.0 - 2022.01.12 初次发布 

    The version of the OpenAPI document: 0.3.4
    Contact: master@mail.paratranz.com
    Generated by OpenAPI Generator (https://openapi-generator.tech)

    Do not edit the class manually.
"""  # noqa: E501


import io
import json
import re
import ssl

import aiohttp
import aiohttp_retry

from paratranz_client.exceptions import ApiException, ApiValueError

RESTResponseType = aiohttp.ClientResponse

ALLOW_RETRY_METHODS = frozenset({'DELETE', 'GET', 'HEAD', 'OPTIONS', 'PUT', 'TRACE'})

class RESTResponse(io.IOBase):

    def __init__(self, resp) -> None:
        self.response = resp
        self.status = resp.status
        self.reason = resp.reason
        self.data = None

    async def read(self):
        if self.data is None:
            self.data = await self.response.read()
        return self.data

    def getheaders(self):
        """Returns a CIMultiDictProxy of the response headers."""
        return self.response.headers

    def getheader(self, name, default=None):
        """Returns a given response header."""
        return self.response.headers.get(name, default)


class RESTClientObject:

    def __init__(self, configuration) -> None:

        # maxsize is number of requests to host that are allowed in parallel
        maxsize = configuration.connection_pool_maxsize

        ssl_context = ssl.create_default_context(
            cafile=configuration.ssl_ca_cert
        )
        if configuration.cert_file:
            ssl_context.load_cert_chain(
                configuration.cert_file, keyfile=configuration.key_file
            )

        if not configuration.verify_ssl:
            ssl_context.check_hostname = False
            ssl_context.verify_mode = ssl.CERT_NONE

        connector = aiohttp.TCPConnector(
            limit=maxsize,
            ssl=ssl_context
        )

        self.proxy = configuration.proxy
        self.proxy_headers = configuration.proxy_headers

        # https pool manager
        self.pool_manager = aiohttp.ClientSession(
            connector=connector,
            trust_env=True
        )

        retries = configuration.retries
        if retries is not None:
            self.retry_client = aiohttp_retry.RetryClient(
                client_session=self.pool_manager,
                retry_options=aiohttp_retry.ExponentialRetry(
                    attempts=retries,
                    factor=0.0,
                    start_timeout=0.0,
                    max_timeout=120.0
                )
            )
        else:
            self.retry_client = None

    async def close(self):
        await self.pool_manager.close()
        if self.retry_client is not None:
            await self.retry_client.close()

    async def request(
        self,
        method,
        url,
        headers=None,
        body=None,
        post_params=None,
        _request_timeout=None
    ):
        """Execute request

        :param method: http request method
        :param url: http request url
        :param headers: http request headers
        :param body: request json body, for `application/json`
        :param post_params: request post parameters,
                            `application/x-www-form-urlencoded`
                            and `multipart/form-data`
        :param _request_timeout: timeout setting for this request. If one
                                 number provided, it will be total request
                                 timeout. It can also be a pair (tuple) of
                                 (connection, read) timeouts.
        """
        method = method.upper()
        assert method in [
            'GET',
            'HEAD',
            'DELETE',
            'POST',
            'PUT',
            'PATCH',
            'OPTIONS'
        ]

        if post_params and body:
            raise ApiValueError(
                "body parameter cannot be used with post_params parameter."
            )

        post_params = post_params or {}
        headers = headers or {}
        # url already contains the URL query string
        timeout = _request_timeout or 5 * 60

        if 'Content-Type' not in headers:
            headers['Content-Type'] = 'application/json'

        args = {
            "method": method,
            "url": url,
            "timeout": timeout,
            "headers": headers
        }

        if self.proxy:
            args["proxy"] = self.proxy
        if self.proxy_headers:
            args["proxy_headers"] = self.proxy_headers

        # For `POST`, `PUT`, `PATCH`, `OPTIONS`, `DELETE`
        if method in ['POST', 'PUT', 'PATCH', 'OPTIONS', 'DELETE']:
            if re.search('json', headers['Content-Type'], re.IGNORECASE):
                if body is not None:
                    body = json.dumps(body)
                args["data"] = body
            elif headers['Content-Type'] == 'application/x-www-form-urlencoded':
                args["data"] = aiohttp.FormData(post_params)
            elif headers['Content-Type'] == 'multipart/form-data':
                # must del headers['Content-Type'], or the correct
                # Content-Type which generated by aiohttp
                del headers['Content-Type']
                data = aiohttp.FormData()
                for param in post_params:
                    k, v = param
                    if isinstance(v, tuple) and len(v) == 3:
                        data.add_field(
                            k,
                            value=v[1],
                            filename=v[0],
                            content_type=v[2]
                        )
                    else:
                        data.add_field(k, v)
                args["data"] = data

            # Pass a `bytes` parameter directly in the body to support
            # other content types than Json when `body` argument is provided
            # in serialized form
            elif isinstance(body, bytes):
                args["data"] = body
            else:
                # Cannot generate the request from given parameters
                msg = """Cannot prepare a request message for provided
                         arguments. Please check that your arguments match
                         declared content type."""
                raise ApiException(status=0, reason=msg)

        if self.retry_client is not None and method in ALLOW_RETRY_METHODS:
            pool_manager = self.retry_client
        else:
            pool_manager = self.pool_manager

        r = await pool_manager.request(**args)

        return RESTResponse(r)





